### PR TITLE
Introduce limit to nested crossref depth in queries

### PR DESCRIPTION
**Pull request by parkerduckworth**
### What's being changed:

With this addition of `QUERY_CROSS_REFERENCE_DEPTH_LIMIT`, Weaviate can now be configured to avoid potential DDoS attacks that attempt to overload the server with queries containing deeply nested cross references.

The depth limit is upheld by a recursive iteration over the properties tree, and includes a short-circuit which is tripped as soon as the max depth is reached.

If `QUERY_CROSS_REFERENCE_DEPTH_LIMIT` is not set, a sane value of `3` is applied by default.

Additionally, with this depth check happening at the `traverser` level, the limit applies to both GQL and gRPC requests.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: usecases/config/config_handler.go [hunk 1]
```diff
@@ -58,6 +58,9 @@ const (
 	DefaultBM25b  = float32(0.75)
 )
 
+// DefaultQueryMaxRefDepth describes the max depth of nested references in a query
+const DefaultQueryMaxRefDepth = 3
```
Comment (antas-marcin) [inline usecases/config/config_handler.go:62]: hmm, shouldn't this setting be defined like [QueryMaximumResults](https://github.com/weaviate/weaviate/blob/main/usecases/config/config_handler.go#L96) and be a part of Config struct?

Comment (antas-marcin) [reply usecases/config/config_handler.go:62]: in traverser you have access to this Config struct like so: `t.config.Config.QueryMaximumResults`

Comment (parkerduckworth) [reply usecases/config/config_handler.go:62]: I agree that would be preferred, and that was also my first thought. But the db config isn't available until much later down the line:
`Traverser` -> `Explorer` -> `objectsSearcher`. 

And then `objectsSearcher` has like 8 different entrypoints into the `DB`:

```go
type objectsSearcher interface {
	hybridSearcher

	// GraphQL Get{} queries
	Search(ctx context.Context, params dto.GetParams) ([]search.Result, error)
	VectorSearch(ctx context.Context, params dto.GetParams) ([]search.Result, error)

	// GraphQL Explore{} queries
	CrossClassVectorSearch(ctx context.Context, vector []float32, targetVector string, offset, limit int,
		filters *filters.LocalFilter) ([]search.Result, error)

	// Near-params searcher
	Object(ctx context.Context, className string, id strfmt.UUID,
		props search.SelectProperties, additional additional.Properties,
		properties *additional.ReplicationProperties, tenant string) (*search.Result, error)
	ObjectsByID(ctx context.Context, id strfmt.UUID, props search.SelectProperties, additional additional.Properties, tenant string) (search.Results, error)
}

type hybridSearcher interface {
	SparseObjectSearch(ctx context.Context, params dto.GetParams) ([]*storobj.Object, []float32, error)
	DenseObjectSearch(context.Context, string, []float32, string, int, int,
		*filters.LocalFilter, additional.Properties, string) ([]*storobj.Object, []float32, error)
	ResolveReferences(ctx context.Context, objs search.Results, props search.SelectProperties,
		groupBy *searchparams.GroupBy, additional additional.Properties, tenant string) (search.Results, error)
}
```

I was hoping to handle this case as early on as possible, in a simple way that won't miss any edge cases

Comment (parkerduckworth) [reply usecases/config/config_handler.go:62]: Great point, will make the change

File: usecases/config/environment.go [hunk 2]
```diff
@@ -254,6 +254,18 @@ func FromEnv(config *Config) error {
 		config.QueryNestedCrossReferenceLimit = DefaultQueryNestedCrossReferenceLimit
 	}
 
+	if v := os.Getenv("QUERY_CROSS_REFERENCE_DEPTH_LIMIT"); v != "" {
+		limit, err := strconv.ParseInt(v, 10, 64)
+		if err != nil {
+			return fmt.Errorf("parse QUERY_CROSS_REFERENCE_DEPTH_LIMIT as int: %w", err)
+		} else if limit <= 0 {
+			limit = math.MaxInt
```
Comment (antas-marcin) [inline usecases/config/environment.go:262]: if limit < 0 then we set the `limit` to `MaxInt`? shouldn't we always default to `DefaultQueryCrossReferenceDepthLimit`?

File: usecases/config/environment.go [hunk 3]
```diff
@@ -254,6 +254,18 @@ func FromEnv(config *Config) error {
 		config.QueryNestedCrossReferenceLimit = DefaultQueryNestedCrossReferenceLimit
 	}
 
+	if v := os.Getenv("QUERY_CROSS_REFERENCE_DEPTH_LIMIT"); v != "" {
+		limit, err := strconv.ParseInt(v, 10, 64)
+		if err != nil {
+			return fmt.Errorf("parse QUERY_CROSS_REFERENCE_DEPTH_LIMIT as int: %w", err)
+		} else if limit <= 0 {
+			limit = DefaultQueryCrossReferenceDepthLimit
```
Comment (antas-marcin) [inline usecases/config/environment.go:262]: or you can return an error like we do it in other cases:
```
return fmt.Errorf("%s must be a positive value larger 0", varName)
```

Comment (antas-marcin) [reply usecases/config/environment.go:262]: oh, we have a method `parsePositiveInt` for parsing int values, maybe we should use it?

**Comment by spiros-spiros** [comment]
Hey team, just wanted to say thanks for turning this around quickly - this is great work.

**Review by antas-marcin — APPROVED** [review]

**Pull request merged**